src/validationinterface.cpp

```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
     boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
 
     CScheduler *scheduler = NULL;
+
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
+    std::list<std::function<void (void)>> callbacksPending;
+    bool fCallbacksRunning = false;
+
+    void MaybeScheduleProcessQueue() {
+        {
+            LOCK(cs_callbacksPending);
+            // Try to avoid scheduling too many copies here, but if we
```

Comment:
> @ryanofsky I believe the point is to avoid duplicate calls to the scheduler since it may be multi-threaded.

I know, this is why the second half of my suggestion was "condition the AddToProcessQueue schedule() call on the queue being previously empty." Anyway, I don't think Matt's particularly interested in these simplifications, and it's easier to communicate these changes as patches rather than english descriptions, so I'd rather just leave any simplifications to followup PRs.

> Still, the code does not guarantee single threaded execution in its current form due to the fCallbacksRunning state being set AFTER the call to schedule()

The reason it works in it's current form is because of the `if (fCallbacksRunning) return;` line at the top of `ProcessQueue()`

Again, I don't think the code in it's current form is the simplest it could be, but it seems safe and easy to clean up later in a followup PR. Also this whole discussion really should be moved to #10179. #10286 is only building on the changes in #10179.